File: go/internal/app/service_test.go
package app

import (
	"context"
	"testing"
	"time"
)

func TestServiceIngestPublishesEvent(t *testing.T) {
	store := NewInMemoryStore()
	pubsub := NewPubSub()
	service := NewService(store, pubsub)

	events, cancel := pubsub.Subscribe(1)
	defer cancel()

	ctx := context.Background()
	takenAt := time.Now().Add(-2 * time.Minute)
	stored, err := service.IngestVital(ctx, "patient-1", 120, 80, takenAt)
	if err != nil {
		t.Fatalf("ingest failed: %v", err)
	}
	if stored.ID == 0 {
		t.Fatal("expected stored vital to have an ID")
	}

	vitals, err := store.ListVitals(ctx)
	if err != nil {
		t.Fatalf("list vitals failed: %v", err)
	}
	if len(vitals) != 1 {
		t.Fatalf("expected 1 vital, got %d", len(vitals))
	}
	if vitals[0].PatientID != "patient-1" {
		t.Fatalf("unexpected patient id: %s", vitals[0].PatientID)
	}

	select {
	case event := <-events:
		if event.Type != EventTypeVitalReceived {
			t.Fatalf("unexpected event type: %s", event.Type)
		}
		if event.Vital.ID != stored.ID {
			t.Fatalf("unexpected vital id: %d", event.Vital.ID)
		}
	case <-time.After(500 * time.Millisecond):
		t.Fatal("timed out waiting for event")
	}
}

func TestServiceIngestValidatesInput(t *testing.T) {
	store := NewInMemoryStore()
	pubsub := NewPubSub()
	service := NewService(store, pubsub)

	ctx := context.Background()
	if _, err := service.IngestVital(ctx, "", 120, 80, time.Now()); err == nil {
		t.Fatal("expected error for missing patient id")
	}
	if _, err := service.IngestVital(ctx, "patient-1", -1, 80, time.Now()); err == nil {
		t.Fatal("expected error for invalid systolic")
	}
	if _, err := service.IngestVital(ctx, "patient-1", 120, 0, time.Now()); err == nil {
		t.Fatal("expected error for invalid diastolic")
	}
	if _, err := service.IngestVital(ctx, "patient-1", 120, 80, time.Time{}); err == nil {
		t.Fatal("expected error for missing taken_at")
	}
}

func TestServiceListAlertsFiltersByPatient(t *testing.T) {
	store := NewInMemoryStore()
	pubsub := NewPubSub()
	service := NewService(store, pubsub)

	ctx := context.Background()
	if _, err := store.AddAlert(ctx, Alert{PatientID: "patient-1", Status: AlertStatusActive}); err != nil {
		t.Fatalf("add alert: %v", err)
	}
	if _, err := store.AddAlert(ctx, Alert{PatientID: "patient-2", Status: AlertStatusActive}); err != nil {
		t.Fatalf("add alert: %v", err)
	}

	alerts, err := service.ListAlerts(ctx, "patient-1")
	if err != nil {
		t.Fatalf("list alerts failed: %v", err)
	}
	if len(alerts) != 1 {
		t.Fatalf("expected 1 alert, got %d", len(alerts))
	}
	if alerts[0].PatientID != "patient-1" {
		t.Fatalf("unexpected patient id: %s", alerts[0].PatientID)
	}
}

func TestServiceListVitalsFiltersByPatient(t *testing.T) {
	store := NewInMemoryStore()
	pubsub := NewPubSub()
	service := NewService(store, pubsub)

	ctx := context.Background()
	if _, err := store.AddVital(ctx, Vital{PatientID: "patient-1"}); err != nil {
		t.Fatalf("add vital: %v", err)
	}
	if _, err := store.AddVital(ctx, Vital{PatientID: "patient-2"}); err != nil {
		t.Fatalf("add vital: %v", err)
	}

	vitals, err := service.ListVitals(ctx, "patient-1")
	if err != nil {
		t.Fatalf("list vitals failed: %v", err)
	}
	if len(vitals) != 1 {
		t.Fatalf("expected 1 vital, got %d", len(vitals))
	}
	if vitals[0].PatientID != "patient-1" {
		t.Fatalf("unexpected patient id: %s", vitals[0].PatientID)
	}
}

File: node/test/service.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import { InMemoryStore } from "../src/domain/store";
import { PubSub } from "../src/domain/pubsub";
import { Service } from "../src/domain/service";
import { AlertStatusActive, EventTypeVitalReceived } from "../src/domain/models";
import { InvalidVitalError, RequestError } from "../src/domain/errors";

const uuidPattern =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

test("service ingest publishes event", async () => {
  const store = new InMemoryStore();
  const pubsub = new PubSub();
  const service = new Service(store, pubsub);

  const { subscription, cancel } = pubsub.subscribe(1);
  const iterator = subscription[Symbol.asyncIterator]();
  const eventPromise = iterator.next();

  const takenAt = new Date(Date.now() - 2 * 60 * 1000);
  const stored = await service.ingestVital(
    "patient-1",
    120,
    80,
    takenAt
  );

  assert.match(stored.id, uuidPattern);
  const vitals = await store.listVitals({ patientId: "patient-1" });
  assert.equal(vitals.items.length, 1);
  assert.equal(vitals.items[0].patientId, "patient-1");

  const result = await eventPromise;
  assert.equal(result.done, false);
  assert.equal(result.value.type, EventTypeVitalReceived);
  assert.equal(result.value.vital.id, stored.id);

  cancel();
});

test("service ingest validates input", async () => {
  const store = new InMemoryStore();
  const pubsub = new PubSub();
  const service = new Service(store, pubsub);

  await assert.rejects(
    service.ingestVital("", 120, 80, new Date()),
    InvalidVitalError
  );
  await assert.rejects(
    service.ingestVital("patient-1", -1, 80, new Date()),
    InvalidVitalError
  );
  await assert.rejects(
    service.ingestVital("patient-1", 120, 0, new Date()),
    InvalidVitalError
  );
  await assert.rejects(
    service.ingestVital("patient-1", 120, 80, new Date("invalid")),
    InvalidVitalError
  );
});

test("service list alerts filters by patient", async () => {
  const store = new InMemoryStore();
  const pubsub = new PubSub();
  const service = new Service(store, pubsub);

  await store.addAlert({ patientId: "patient-1", status: AlertStatusActive });
  await store.addAlert({ patientId: "patient-2", status: AlertStatusActive });

  const alerts = await service.listAlerts({ patientId: "patient-1" });
  assert.equal(alerts.items.length, 1);
  assert.equal(alerts.items[0].patientId, "patient-1");
});

test("service list alerts requires patient id", async () => {
  const store = new InMemoryStore();
  const pubsub = new PubSub();
  const service = new Service(store, pubsub);

  await assert.rejects(service.listAlerts({ patientId: "" }), RequestError);
});

test("service list vitals filters by patient", async () => {
  const store = new InMemoryStore();
  const pubsub = new PubSub();
  const service = new Service(store, pubsub);

  await store.addVital({ patientId: "patient-1" });
  await store.addVital({ patientId: "patient-2" });

  const vitals = await service.listVitals({ patientId: "patient-1" });
  assert.equal(vitals.items.length, 1);
  assert.equal(vitals.items[0].patientId, "patient-1");
});

test("service list vitals requires patient id", async () => {
  const store = new InMemoryStore();
  const pubsub = new PubSub();
  const service = new Service(store, pubsub);

  await assert.rejects(
    service.listVitals({ patientId: "   " }),
    RequestError
  );
});
